pkg/blockbuilder/tsdb.go

```diff
@@ -0,0 +1,398 @@
+	if err := os.RemoveAll(udir); err != nil {
+		return nil, err
+	}
+	if err := os.MkdirAll(udir, os.ModePerm); err != nil {
+		return nil, err
+	}
+
+	userID := tenant.id
+	userLogger := util_log.WithUserID(userID, b.logger)
+
+	udb := &userTSDB{
+		userID: userID,
+	}
+
+	db, err := tsdb.Open(udir, userLogger, nil, &tsdb.Options{
+		RetentionDuration:           0,
+		MinBlockDuration:            2 * time.Hour.Milliseconds(),
+		MaxBlockDuration:            2 * time.Hour.Milliseconds(),
+		NoLockfile:                  true,
+		StripeSize:                  b.config.TSDB.StripeSize,
+		HeadChunksWriteBufferSize:   b.config.TSDB.HeadChunksWriteBufferSize,
+		HeadChunksEndTimeVariance:   b.config.TSDB.HeadChunksEndTimeVariance,
+		HeadChunksWriteQueueSize:    b.config.TSDB.HeadChunksWriteQueueSize,
+		WALSegmentSize:              -1, // No WAL
+		SeriesLifecycleCallback:     udb,
+		BlocksToDelete:              udb.blocksToDelete,
+		IsolationDisabled:           true,
+		EnableOverlappingCompaction: false,                                                // always false since Mimir only uploads lvl 1 compacted blocks
+		OutOfOrderTimeWindow:        b.limits.OutOfOrderTimeWindow(userID).Milliseconds(), // The unit must be same as our timestamps.
+		OutOfOrderCapMax:            int64(b.config.TSDB.OutOfOrderCapacityMax),
+		EnableNativeHistograms:      b.limits.NativeHistogramsIngestionEnabled(userID),
+		SecondaryHashFunction:       nil, // TODO(codesome): May needed when applying limits. Used to determine the owned series by an ingesters
+	}, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	db.DisableCompactions()
+
+	udb.db = db
+
+	return udb, nil
+}
+
+func (b *TSDBBuilder) tsdbDir(tenant tsdbTenant) string {
+	return filepath.Join(b.config.TSDB.Dir, strconv.Itoa(int(tenant.part)), tenant.id)
+}
+
+// CompactAndUpload compacts the blocks of all the TSDBs and uploads them.
+// blockUploaderForUser should give a function to upload the blocks of a particular user.
+func (b *TSDBBuilder) CompactAndUpload(ctx context.Context, blockUploaderForUser func(context.Context, string) blockUploader) (int, error) {
+	b.tsdbsMu.Lock()
+	defer b.tsdbsMu.Unlock()
+
+	level.Info(b.logger).Log("msg", "compacting and uploading blocks", "num_tsdb", len(b.tsdbs))
+
+	if len(b.tsdbs) == 0 {
+		return 0, nil
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
+	if b.config.TSDB.ShipConcurrency > 0 {
+		eg.SetLimit(b.config.TSDB.ShipConcurrency)
+	}
+	numBlocks := 0
+	for tenant, db := range b.tsdbs {
+		if err := db.compactEverything(ctx); err != nil {
+			return numBlocks, err
+		}
+
+		var blockNames []string
+		dbDir := db.db.Dir()
+		for _, b := range db.db.Blocks() {
+			blockNames = append(blockNames, b.Meta().ULID.String())
+		}
+		numBlocks += len(blockNames)
+
+		if err := db.Close(); err != nil {
+			return numBlocks, err
+		}
+
+		delete(b.tsdbs, tenant)
+
+		// Make a copy of userID because 'tenant' changes in the next iteration.
+		userID := tenant.id
```

Comment:
> [..] So you don't have to worry about them anymore.

Since Go 1.23 is out now, I hope we will soon update `go.mod` Go 1.22, which fixed it on [language level][1]. After that we can wipe this piece of Go knowledge for better ;)

[1]: https://go.dev/wiki/LoopvarExperiment#my-test-fails-with-the-change-how-can-i-debug-it